client/src/main/java/com/alibaba/nacos/client/naming/core/HostReactor.java

```diff
@@ -471,9 +472,12 @@ public void run() {
                 
                 lastRefTime = serviceObj.getLastRefTime();
                 
-                if (!notifier.isSubscribed(serviceName, clusters) && !futureMap
-                        .containsKey(ServiceInfo.getKey(serviceName, clusters))) {
+                if (!notifier.isSubscribed(serviceName, clusters)) {
```

Comment:
Why remove futureMap.containsKey?

Reply:
I searched for the usage of futureMap. There are only get and put operations. Once a service is put into futureMap, it can never be removed. This will make futureMap.containsKey() always return true and make the if block can not be entered.

Reply:
If judgement like this, It will cause problem when I call NamingService.getAllIsntance(subscribe = true).

The cache in HostReactor is always old one, will not be refresh.

Reply:
If the cache refers to serviceInfoMap，is it OK to clear the cache after the judgement? Like HostReactor.java Line 479 in the PR.

Reply:
If you real want to remove cache. Maybe you clear it when you call unsubscribe not in HostReactor.

Reply:
When unsubscribe is called, the next round UpdateTask may be already scheduled. If the cache is cleared at this moment, the next round UpdateTask will fill the cache again.
I think it's better to clear the cache exactly when the UpdateTask can be terminated, to make sure the UpdateTask will not run afterwards.

Reply:
> If judgement like this, It will cause problem when I call NamingService.getAllIsntance(subscribe = true).

> The cache in HostReactor is always old one, will not be refresh.

can't solve this problem. It will cause data no right. So I think only do once more query is no much problem.

Reply:
I try to summarize the related logic as below.

1. NamingService.getAllIsntance(subscribe = true) reads cache data from HostReactor.serviceInfoMap. If serviceInfoMap has no data that belongs to a service, there is no old data problem.

2. When unsubscribe clears all listeners in HostReactor.InstancesChangeNotifier, it indicates that the UpdateTask can be stoped. The original judgement `futureMap.containsKey()` has no effect, it always returns true and should be removed.

3. When UpdateTask can be judged to be stopped, change the flag stop = true so that no more UpdateTask will run and no more cache filling operation can be executed. Then clear the cache in serviceInfoMap and make no old data problem caused. The key codes are as follows
```
if (!notifier.isSubscribed(serviceName, clusters)) {
    // abort the update task
    stop = true;
    String serviceKey = ServiceInfo.getKey(serviceName, clusters);
    serviceInfoMap.remove(serviceKey);
    futureMap.remove(serviceKey);
    return;
}
``` 

4. If cache is cleared after unsubscribe will case problem. Suppose that 
10:00:01 unsubscribe is called, serviceInfoMap is cleared.
10:00:02 UpdateTask runs, fills the cache before the judgement by 
```
ServiceInfo serviceObj = serviceInfoMap.get(ServiceInfo.getKey(serviceName, clusters));
                
if (serviceObj == null) {
    updateService(serviceName, clusters);
    return;
}
```
Afterwards the task is stopped after judgement without clearing the cache, the old data will remain in serviceInfoMap.

It would be appreciated if you tell me which of the 4 points is wrong. Then I can better understand where the problem is. Thanks.

Reply:
> I try to summarize the related logic as below.
> 
> 1. NamingService.getAllIsntance(subscribe = true) reads cache data from HostReactor.serviceInfoMap. If serviceInfoMap has no data that belongs to a service, there is no old data problem.
> 2. When unsubscribe clears all listeners in HostReactor.InstancesChangeNotifier, it indicates that the UpdateTask can be stoped. The original judgement `futureMap.containsKey()` has no effect, it always returns true and should be removed.
> 3. When UpdateTask can be judged to be stopped, change the flag stop = true so that no more UpdateTask will run and no more cache filling operation can be executed. Then clear the cache in serviceInfoMap to avoid old data problem The key codes are as follows
> 
> ```
> if (!notifier.isSubscribed(serviceName, clusters)) {
>     // abort the update task
>     stop = true;
>     String serviceKey = ServiceInfo.getKey(serviceName, clusters);
>     serviceInfoMap.remove(serviceKey);
>     futureMap.remove(serviceKey);
>     return;
> }
> ```
> 
> 4. If cache is cleared after unsubscribe will case problem. Suppose that
>    10:00:01 unsubscribe is called, serviceInfoMap is cleared.
>    10:00:02 UpdateTask runs, fills the cache before the judgement by
> 
> ```
> ServiceInfo serviceObj = serviceInfoMap.get(ServiceInfo.getKey(serviceName, clusters));
>                 
> if (serviceObj == null) {
>     updateService(serviceName, clusters);
>     return;
> }
> ```
> 
> Afterwards the task is stopped after judgement without clearing the cache, the old data will remain in serviceInfoMap.
> 
> It would be appreciated if you tell me which of the 4 points is wrong. Then I can better understand where the problem is. Thanks.

If like this. Why nacos add `NamingService.getAllIsntance(subscribe = true)`? After your changes, `NamingService.getAllIsntance(subscribe = true)` is same as `NamingService.getAllIsntance(subscribe = false)`. And use one more thread and task to do. So I think we can't change like this.

Reply:
After these changes, NamingService.getAllIsntance(subscribe = true) can work as before. It still return instances and start to subscribe. There is no difference before unsubscribe() is called. NamingService.getAllIsntance(subscribe = false) does not start subscribing, these two invocations do not have the same behavior.

But when unsubscribe operation is needed, these changes can make sense. Without these changes, even if a service is unsubscribed and all related listeners are deregistered, the UpdateTask is still running. This will case some (maybe a lot of) useless UpdateTasks run continuously. Both client side and server side have to pay the cost of useless polling. This also breaks the definition of unsubscribe: poling is expected to be stopped after unsubscribe in common sense.

I think it is a problem worth to be fixed. If these changes are not good enough, other fix plans are still recommended. If this problem can be ignored or it is intentionally designed like that, it's also OK to me.

Reply:
`NamingService.getAllIsntance(subscribe = true)` no add subscriber to notify. So the first time updateTask run will remove cache and task after your change.

Reply:
Yes, you are right. Now I know where the problem is. Thanks a lot for explaining that to me. Meanwhile I have changed the pr as follows.

1. Keep the original judgement, use futureMap as a flag to judge whether to stop the update task. Therefore `NamingService.getAllIsntance(subscribe = true)` can work as before. Update task will not stop when there is no listener.
2. Remove the futureMap flag when unsubscribe is called and all listeners are removed in order to fix the issue.

Hope it's a better solution now.